### ISSUE TITLE
Folding/Indenting issues

### ISSUE THREAD

**Issue by smartspot2** [lervag/vimtex#1813]
**Issue**
Folding seems to break in `enumerate` environments when a new environment is started on the same line as an `\item` (perhaps related to the additions in #1765). With the tex file below, there is no fold on the line that says `\item Folding issue`, and the `enumerate` environment fold ends early with the first item (attempting to `zc` after the first item gives `E490: No fold found`).

I've also noticed (probably completely unrelated to the folding issue above) that defining environments inline after an `\end{...}` seems to break indenting for the rest of the document. Running `gg=G` on the tex file below causes the last three lines to be erroneously indented an extra time, when I expect no changes to be made.

**minimal.vim**

```vim
set nocompatible
let &runtimepath  = '~/.vim/plugged/vimtex,' . &runtimepath
let &runtimepath .= ',~/.vim/plugged/vimtex/after'
filetype plugin indent on
syntax enable
let g:vimtex_fold_enabled=1
```

**minimal.tex**

```tex
\documentclass{minimal}
\begin{document}
\begin{enumerate}
    \item $A = \begin{bmatrix}
            1 & 2 \\
            3 & 4
        \end{bmatrix}$

    \item Folding issue
\end{enumerate}

Indenting issue:
\[
    \begin{bmatrix}
        1 & 2 \\
        3 & 4
    \end{bmatrix} \begin{bmatrix} 5 \\ 6 \end{bmatrix}
\]
\end{document}
```

P.S. I also just want to say thank you for all of the work and dedication you've put in to making, maintaining, and improving this plugin—it has played a huge part in streamlining my homework/notetaking process in college at the moment.

**Comment by lervag** [comment]
> **Issue**
> Folding seems to break in `enumerate` environments when a new environment is started on the same line as an `\item` (perhaps related to the additions in #1765). With the tex file below, there is no fold on the line that says `\item Folding issue`, and the `enumerate` environment fold ends early with the first item (attempting to `zc` after the first item gives `E490: No fold found`).

Thanks, I can reproduce this. I'll look into it!

> I've also noticed (probably completely unrelated to the folding issue above) that defining environments inline after an `\end{...}` seems to break indenting for the rest of the document. Running `gg=G` on the tex file below causes the last three lines to be erroneously indented an extra time, when I expect no changes to be made.

I have two comments here:

1. It makes my life easier if we keep each issue/thread focussed on a single issue. Thus, the indentation problem should be opened as a separate issue.

2. I can confirm that the issue is reproducible, _but_: You're making my life hard (*jokingly*). What I mean is, it is hard to parse LaTeX code due to its semi unstructured nature. The current implementation works in most cases, but there are edge cases. I think it could be possible to fix the issue, but I'm wondering if it is necessary. That is, your LaTeX source code would be better (i.e. more readable) if you added a newline after the first `\end{bmatrix}`, and then the indents will also be properly parsed. If you still feel that I should fix this, then open a new issue for it - the sample LaTeX code, a minimal description, and a reference to this thread would suffice as input.

> P.S. I also just want to say thank you for all of the work and dedication you've put in to making, maintaining, and improving this plugin—it has played a huge part in streamlining my homework/notetaking process in college at the moment.


Thanks! I'm happy to hear my work brings value to others! :)

**Comment by lervag** [comment]
> Thanks, I can reproduce this. I'll look into it!

I think I fixed it; can't promise it is watertight, but it seems to work.
